### PR TITLE
Bug 389815 - Enhancement Request - JSON specific multidimensional array support

### PR DESCRIPTION
**Pull request by rfelcman**
This is fix/implementation for this request. It is fix for unmarshaler, marshaller + unit tests.
There is new marshaller property `MarshallerProperties.JSON_DISABLE_NESTED_ARRAY_NAME` `true|false` to ensure backward compatibility. Why:

Before this fix (or with `MarshallerProperties.JSON_DISABLE_NESTED_ARRAY_NAME = false` (default value)) MOXy marshalled values from nested array fields
`@XmlElement(name="char2dArray")
private char[][] char2dArray;`
into `"char2dArray":[{"item":["a","b","c","d"]},{"item":["e","f","g","h"]}]`

After this fix (or with `MarshallerProperties.JSON_DISABLE_NESTED_ARRAY_NAME = true`) output will be: `"char2dArray": [["a","b","c","d"],["e","f","g","h"]`

Of course unmarshaller will process multi-dimensional arrays too.

### CODE REVIEW

**Comment by raner** [comment]
Nice work, Radek!

Once this gets merged I can finally switch back from Jackson to MOXy 👍

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonBuilderRecord.java [hunk 1]
```diff
@@ -197,9 +197,9 @@ protected void addValueToArray(Level level, Object value, QName schemaType){
         private JsonObjectBuilder jsonObjectBuilder;
         private JsonArrayBuilder jsonArrayBuilder;
 
-        public Level(boolean isCollection, Level position) {
-            super(isCollection, position);
-        }
+         public Level(boolean isCollection, Level position, boolean nestedArray) {
```
Comment (michael-o) [inline foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonBuilderRecord.java]: Leading spaces...

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 // Contributors:
 //     Denise Smith - 2.6 - initial implementation
+//     Radek Felman - 2.7.5 - Bug 389815 - Enhancement Request - JSON specific multidimensional array support
```
Comment (michael-o) [inline foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java]: I think something isn't necessary in the code becuase the credits are in the git commit.

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java [hunk 3]
```diff
@@ -170,20 +189,23 @@ public void endDocument() {
...
     public void startCollection() {
         if(position == null){
              isRootArray = true;
-             position = createNewLevel(true, null);
+             position = createNewLevel(true, null, false);
              startRootLevelCollection();
         } else {
             if(isLastEventStart){
-                setComplex(position, true);
+                    setComplex(position, true);
```
Comment (michael-o) [inline foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java]: Indent?

File: moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java [hunk 4]
```diff
@@ -205,6 +205,43 @@
...
+     * ...
+     * "char2dArray": [
+     *      ["a","b","c","d"],
+     *      ["e","f","g","h"]
+     * ]
+     * ...
+     * </pre>
+     * @since 2.7.5
+     */
+    public static final String JSON_DISABLE_NESTED_ARRAY_NAME = "eclipselink.json.disable-nested-array-name";
```
Comment (michael-o) [inline moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java]: Why not use the JAXB Property instead of a literal?

File: moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/compiler/AnnotationsProcessor.java [hunk 5]
```diff
@@ -195,12 +195,12 @@
 public final class AnnotationsProcessor {
     static final String JAVAX_ACTIVATION_DATAHANDLER = "javax.activation.DataHandler";
     static final String JAVAX_MAIL_INTERNET_MIMEMULTIPART = "javax.mail.internet.MimeMultipart";
+    static final String ARRAY_PACKAGE_NAME = "jaxb.dev.java.net.array";
     private static final String JAVAX_XML_BIND_JAXBELEMENT = "javax.xml.bind.JAXBElement";
     private static final String JAVAX_XML_BIND_ANNOTATION = "javax.xml.bind.annotation";
     private static final String OXM_ANNOTATIONS = "org.eclipse.persistence.oxm.annotations";
     private static final String TYPE_METHOD_NAME = "type";
     private static final String VALUE_METHOD_NAME = "value";
-    private static final String ARRAY_PACKAGE_NAME = "jaxb.dev.java.net.array";
```
Comment (michael-o) [inline moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/compiler/AnnotationsProcessor.java]: Why did you change the position of this static field?

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java (hunk 2 above)
Comment (rfelcman) [reply foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java]: This is not about credit, but there are so many changes so just keep there notice.

File: moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java (hunk 4 above)
Comment (rfelcman) [reply moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java]: But this property is used only in marshaller scope. JAXB context property means booth unmarshaller and marshaller.

Comment (michael-o) [reply moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java]: Alright, the unmarshaller will properly handle this with or without the property?

Comment (rfelcman) [reply moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java]: Yes unmarshaller will handle it automatically without any additional properties.

Comment (michael-o) [reply moxy/org.eclipse.persistence.moxy/src/org/eclipse/persistence/jaxb/MarshallerProperties.java]: This is then fine by me.

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/oxm/record/json/JsonStructureReader.java [hunk 6]
```diff
@@ -28,15 +28,7 @@
...
-import javax.json.Json;
-import javax.json.JsonArray;
-import javax.json.JsonNumber;
-import javax.json.JsonException;
-import javax.json.JsonObject;
-import javax.json.JsonReader;
-import javax.json.JsonString;
-import javax.json.JsonStructure;
-import javax.json.JsonValue;
+import javax.json.*;
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/internal/oxm/record/json/JsonStructureReader.java]: I don't like this... use FQN imports

File: foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java [hunk 7]
```diff
@@ -674,15 +696,37 @@ public void setNil(boolean isNil) {}
 
         protected boolean isCollection;
         protected boolean emptyCollection;
+         protected boolean emptyCollectionGenerated;
```
Comment (lukasj) [inline foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java]: there seems to be offending whitespace here

Comment (lukasj) [reply foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/oxm/record/JsonRecord.java]: ...make sure formatting is consistent through the file

**Comment by lukasj** [comment]
@rfelcman rebase this, so this can be merged cleanly, please. Thanks!
